src/vivarium_testing_utils/automated_validation/data_transformation/calculations.py

```diff
@@ -131,3 +131,58 @@ def get_singular_indices(data: pd.DataFrame) -> dict[str, Any]:
         if data.index.get_level_values(index_level).nunique() == 1:
             singular_metadata[index_level] = data.index.get_level_values(index_level)[0]
     return singular_metadata
+
+
+def weighted_average(
+    data: pd.DataFrame, weights: pd.DataFrame, stratifications: list[str] = []
+) -> pd.Series[float | int]:
+    """Calculate a weighted average of the data using the provided weights.
+
+    Parameters
+    ----------
+    data
+        DataFrame with the values to average.
+    weights
+        DataFrame with the weights to apply to the values in data.
+    stratifications
+        List of column names to use for stratification.
+
+    Returns
+    -------
+        Pandas series with the weighted average values. If the index of data and weights have
+        different levels, the function will boradcast the weighted average. For example, if
+        the following data and weights are provided:
+        data = pd.DataFrame({"value": [10, 20, 30, 40]}, index=pd.MultiIndex.from_tuples(
+            ("location_1", "sex_1", "age_1"),
+            ("location_1", "sex_1", "age_2"),
+            ("location_1", "sex_2", "age_1"),
+            ("location_1", "sex_2", "age_2"),
+        ))
+        weights = pd.DataFrame({"value": [1, 2, 3, 4]}, index=pd.MultiIndex.from_tuples(
+            ("location_1", "sex_1")
+            ("location_1", "sex_1"),
+            ("location_1", "sex_2"),
+            ("location_1", "sex_2"),
+        ))
+        a pandas series would still be returned where the value column will just be broadcasted
+        across the stratifications.
+
+    """
+    # Check that index levels are compatible (at least subsets of each other)
+    data_index_names = set(data.index.names)
+    weights_index_names = set(weights.index.names)
+
+    if not (
+        data_index_names.issubset(weights_index_names)
+        or weights_index_names.issubset(data_index_names)
+    ):
+        raise ValueError(
```

Comment:
I am doing this because if an index is present in one but not the other, a dataframe would be returned instead where the index level(s) that do not exist in both data and weights would be all NaNs and so I think we should avoid that behavior but if the indexes are equal or subsets of each other then it is fine to broadcast them.